### PR TITLE
interpret allowed updates and other actions in a free Trigger monad, like Script

### LINKED ISSUE #7372: Monadifying the low-level trigger API
Currently, the `update` function in the trigger API is a simple function that returns the new state and a list of commands: https://github.com/digital-asset/daml/blob/f67af49044a4a4df24ecf01247cfbceada98751f/triggers/daml/Daml/Trigger/LowLevel.daml#L154

This means that we cannot hand over control to the trigger runner during the update. There are a few issues as a consequence of this:

1. We cannot generate random UUIDs when the user calls `emitCommands` and return them as a result (DAML has no random number generation). Currently we work around this by using an incrementing command id in DAML API and mapping that to random UUIDs on the Scala side. However, that is both messy and it makes it harder to debug triggers as you don’t easily get access to the command id used to submit a list of commands.

2. To add exception handling to triggers (matching the one we want to add for `Update`s and DAML Script), we need special LF primitives. This goes against the current design where triggers deliberately are not LF builtins.

3. While we provide `emitCommands` in the high-level trigger API, the commands are only sent at the end of the rule rather than immediately. This has potential for confusion (consider the case where your trigger crashes after calling `emitCommands` without those commands ever being sent). However, I haven’t seen anyone run into this so far so I think it’s a relatively small issue.

4. The API makes it hard to add new effectful features. For example, consider adding logging functionality. Our best option here is either to rely on `trace` which is notoriously confusing since it’s tied to evaluation and not execution or return a list of log messages as part of each update which is confusing since they are delayed.

1 and 2 are issues that I expect we will have to solve fairly soon. 3 and 4 are more potential issues that could cause issues down the line. Right now, we don’t have concrete plans for adding more effects but at least (user-defined) logging will almost certainly come up at some point.

DAML Script does address this by building on top of a free monad. This allows us to hand over control to the Scala side at any point to implement effects. For example, `submit` is handled that way. Not quite a free-monad but still possible with the same design is exception handling. See https://github.com/digital-asset/daml/pull/7368 for a prototype of how this could work without requiring special catching primitives in LF.

### PR DESCRIPTION
**Pull request by S11001001**
Fixes #7372.

```rst
CHANGELOG_BEGIN
- [Triggers] The ``Time`` argument was removed from the trigger rule function; instead, it
  can be fetched within the ``TriggerA`` ``do`` block by ``getTime``, as with ``Update``
  and ``Scenario``.  The ``LowLevel`` trigger interface has been redesigned; such triggers
  need to be rewritten or ported to high-level triggers.
  See `issue #7456 <https://github.com/digital-asset/daml/pull/7456>`_.
CHANGELOG_END
```

This ended up incorporating the high-level Trigger.rule as well as the low-level trigger interface. We can revert that part if it is undesirable.

* [x] all command submits ultimately expressed as TriggerF actions
* [x] reworked low-level triggers pass integration tests
* [x] high-level TriggerA incorporates whole TriggerF algebra
* [x] Time argument removed
* [x] submissions moved earlier, to emitCommands?
* [ ] different results from `getTime` within rule?
* [x] transform rule output to Script monad for testRule (librifying Script's Free)
* [x] reexport more LowLevel symbols from high-level module (no more were needed)
* [x] delete comments from daml-script/converter

Compile-wise, the only change a high-level trigger author will see immediately is that the time argument is replaced by the `getTime` action in `TriggerA`, which is exactly the same `getTime` as in `Update` et al. (We may consider giving `getTime` more semantically complex behavior than "just like the time argument from before".)

A low-level trigger author can simply use the `forwardPort*` functions as used in the tests herein to convert their old functions to the new format.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 1]
```diff
@@ -244,3 +255,111 @@ data Commands = Commands
...
+  Free x >>= f = Free ((>>= f) <$> x)
+
+liftFC : Functor f => f a -> Free f a
+liftFC fa = Free (Pure <$> fa)
+
+lowerFC : Action f => Free f a -> f a
+lowerFC (Pure a) = pure a
+lowerFC (Free x) = x >>= lowerFC
+
+newtype StateT s m a = StateT { runStateT : s -> m (a, s) }
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:285]: Note that newtypes are not free in DAML so you are sometimes better off inlining this stuff (sadly). Not too worried about this for triggers for now but worth keeping in mind.

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 2]
```diff
@@ -244,3 +255,111 @@ data Commands = Commands
...
+execStateT (StateT fa) = fmap snd . fa
+
+zoom : Functor m => (t -> s) -> (t -> s -> t) -> StateT s m a -> StateT t m a
+zoom r w (StateT smas) = StateT $ \t ->
+  smas (r t) <&> \(a, s) -> (a, w t s)
+
+-- Must be kept in sync with Runner#handleStepFreeResult
+data TriggerF a =
+  GetTime (Time -> a)
+  | Submit (Commands, () -> a)
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:321]: As mentioned on the issue, one thing that we can do with this monadified version is to return a `CommandId` here so we don’t have to do this annoying back and forth translation between the random UUIDs in the trigger runner and what users specify here. Let’s leave that for a separate PR though.

File: triggers/daml/Daml/Trigger/Internal.daml [hunk 3]
```diff
@@ -104,23 +103,28 @@ applyTransaction : Transaction -> ACS -> ACS
...
+  : (Party -> ACS -> Map CommandId [Command] -> s -> TriggerA ())
+  -> TriggerRule (TriggerState s) ()
+runRule rule = do
+  state <- get
+  cmds <- TriggerRule . zoom zoomIn zoomOut $ do
+    runTriggerRule . runTriggerA
+        $ rule state.party state.acs state.commandsInFlight state.userState
+    aState <- get
+    pure aState.emittedCommands
+  mapA_ submitCommands cmds -- TODO don't procrastinate
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/Internal.daml:115]: I’m fine with leaving this for a separate PR if you want.

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 1 above)
Comment (cocreature) [reply triggers/daml/Daml/Trigger/LowLevel.daml:285]: Also I just realized that we used a newtype before as well … Maybe at least worth merging it to one layer of newtypes at some point instead of having `TriggerA` on top of `TriggerRule` on top of `StateT` …. Doesn’t have to be in this PR though.

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 4]
```diff
@@ -244,3 +255,111 @@ data Commands = Commands
...
+instance HasTime (TriggerRule s) where
+  getTime = liftTF (GetTime identity)
+
+submitCommands : ActionTrigger m => Commands -> m ()
+submitCommands cmds = liftTF (Submit (cmds, identity))
+
+type TriggerInitialState s = Party -> Time -> ActiveContracts -> (s, [Commands])
+type TriggerUpdate s = Time -> Message -> s -> (s, [Commands])
+
+forwardPortInitialState : TriggerInitialState s -> Party -> ActiveContracts -> Free TriggerF s
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:353]: Not that worried about trying to make it easy to migrate the low-level trigger API. Triggers are still alpha in general and the low-level API is even more likely to change.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 5]
```diff
@@ -217,53 +220,70 @@ class Runner(
...
+  private def handleCommands[Z](
+      commandId: String,
+      commands: Seq[Command],
+      submit: SubmitRequest => Z): Z = {
+    if (usedCommandIds.contains(commandId)) {
+      throw new RuntimeException(s"Duplicate command id: $commandId")
+    }
+    usedCommandIds += commandId
+    val commandUUID = UUID.randomUUID
+    commandIdMap += (commandUUID -> commandId)
```
Comment (cocreature) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:235]: That’s the mapping I was talking about above which we can get rid off with this change.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 6]
```diff
@@ -217,53 +220,70 @@ class Runner(
...
-        throw new RuntimeException(s"Expected Tuple2 but got $v")
+            case _ => {
+              case _ =>
+                val msg = s"unrecognized TriggerF step $variant"
+                logger.error(msg)
+                throw new ConverterException(msg)
+            }
+          }(fallback = throw new ConverterException(s"invalid contents for $variant: $vv"))
+        case Right(Left(newState)) => newState
+        case Left(e) => throw new RuntimeException(e)
```
Comment (cocreature) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:283]: This was here before but a `ConverterException` might be more appropriate here.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 7]
```diff
@@ -497,6 +519,21 @@ object Runner extends StrictLogging {
     }
   }
 
+  private[this] val DaTypesTuple2 =
```
Comment (cocreature) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:522]: Maybe a better fit for the converter? Probably even the shared converter.

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 8]
```diff
@@ -244,3 +255,111 @@ data Commands = Commands
...
+  Free x <*> my     = Free ((<*> my) <$> x)
+
+instance Functor f => Action (Free f) where
+  Pure a >>= f = f a
+  Free x >>= f = Free ((>>= f) <$> x)
+
+liftFC : Functor f => f a -> Free f a
+liftFC fa = Free (Pure <$> fa)
+
+lowerFC : Action f => Free f a -> f a
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:281]: Seems to be unused.

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 9]
```diff
@@ -155,8 +166,8 @@ data ActiveContracts = ActiveContracts { activeContracts : [Created] }
 -- | Trigger is (approximately) a left-fold over `Message` with
 -- an accumulator of type `s`.
 data Trigger s = Trigger
-  { initialState : Party -> Time -> ActiveContracts -> (s, [Commands])
-  , update : Time -> Message -> s -> (s, [Commands])
+  { initialState : Party -> ActiveContracts -> Free TriggerF s
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:169]: While this is only the LowLevel interface, some newtype around `Free TriggerF s` would probably still be nice so that `submitCommands` can also be used here if you really want this interface with all the caveats that come with it.

**Review by cocreature — COMMENTED** [review]
Very nice, Thanks a lot! This is pretty much exactly what I had in mind.

One thing to note: You will get some failures in the compatibility tests. We currently test that a trigger runner can run DARs from an older SDK. With triggers being alpha, this isn’t something that we have to support (and we never said we do support it) so I think the most reasonable option here is probably to modify the tests to exclude the version combinations that will be broken by this. Happy to help you find the right place to set that up.

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 9 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml:169]: It's already possible to do this by writing `runTriggerSetup do`.

Comment (cocreature) [reply triggers/daml/Daml/Trigger/LowLevel.daml:169]: good point, you still end up leaking `Free` into the public API though which I’d prefer to avoid.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 6 above)
Comment (S11001001) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: de429c7fa535509ed61b4ef0f98b1d6c31da786a

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 7 above)
Comment (S11001001) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: f8285a579128d7151ff30a661b5241dc4b87a9b8

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 8 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml]: 3669c03e40041958b76d57c96ef794edcdcb4cc8

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 4 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml:339]: This was more for my own sanity-checking (that all ports must be possible) and my own ease of porting the integration tests. If it seems nicer to directly port the tests I'll get rid of them.

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 9 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml]: I was avoiding this because (as you mentioned in another comment) newtypes add layers, which is annoying to deal with at the Scala layer as seen in 841ee75320120e2300d34bf2ec563e17299029af .  Anyway that is there now.

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 1 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml:263]: I use them for three things in this PR:

1. matching constraint signatures nicely,
2. letting the functor layer go up and down when I haven't settled on which layer will operate where. e.g. TriggerSetup very nearly got elided for the final design, but it was needed for the low-level initialState as you wanted, so it survived causing very little trouble at the DAML layer (at the Scala layer is another story), and
3. _not_ letting operations pollute upwards. For example it would be very undesirable for `ActionState TriggerAState TriggerA` to exist; it might even be desirable to remove `ActionTrigger TriggerA` so as to hide `submitCommands` from the high-level API.

(2) is naturally just scaffolding, but we should take care about (3). And of course damlc will just yell at you for not thinking about (1).

File: triggers/daml/Daml/Trigger/Internal.daml (hunk 3 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/Internal.daml]: Nah, we'll do it live. 0ea3d9c894

File: compatibility/BUILD [hunk 10]
```diff
@@ -79,9 +79,12 @@ config_setting(
     if versions.is_at_least(sdk_version, platform_version)
 ]
 
+last_pre_7456_trigger_version = "1.6.0-snapshot.20200930.5312.0.b9a1905d"
+
 [
     daml_trigger_dar(sdk_version)
     for sdk_version in sdk_versions
+    if not versions.is_at_most(last_pre_7456_trigger_version, sdk_version)
```
Comment (cocreature) [inline compatibility/BUILD:87]: This seems a bit too strong. We don’t want to kill those compat tests for newer versions completely. What we can remove is testing a version > last_pre_7456_trigger_version against a DAR built with an older version.

File: daml-script/daml/Daml/Script/Free.daml [hunk 11]
```diff
@@ -0,0 +1,34 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Daml.Script.Free
```
Comment (cocreature) [inline daml-script/daml/Daml/Script/Free.daml:4]:
```suggestion
-- | HIDE
module Daml.Script.Free
```

Given that we deliberately hide this from the public API, we should also hide it from the docs. We could also follow the pattern that we use in the trigger library and add an `Internal` in the module name to make it very explicit that this is not part of the supported API.

File: triggers/daml/Daml/Trigger.daml [hunk 12]
```diff
@@ -104,6 +103,7 @@ emitCommands cmds pending = do
     , pendingContracts = Map.insert id pending s.pendingContracts
     , nextCommandId = s.nextCommandId + 1
     }
+  submitCommands commands
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger.daml:106]: Not part of this PR (I might have already mentioned this in #7392 not quite sure) but we could consider deprecating the name `emitCommands` in favor of `submitCommands`. `emit` isn’t really terminology that we use anywhere else and just seems confusing. I don’t quite want to use just `submit` since we use that for the synchronous submissions in scenarios and daml script.

File: triggers/daml/Daml/Trigger.daml [hunk 13]
```diff
@@ -184,35 +184,37 @@ runTrigger userTrigger = LowLevel.Trigger
...
   }
   where
-    initialState party time (ActiveContracts createdEvents) =
+    initialState party (ActiveContracts createdEvents) =
       let acs = foldl (\acs created -> applyEvent (CreatedEvent created) acs) (ACS mempty Map.empty) createdEvents
           userState = userTrigger.initialize acs
           state = TriggerState acs party userState Map.empty 0
-      in runRule userTrigger.rule time state
-    update time msg state =
+      in TriggerSetup $ execStateT (runTriggerRule $ runRule userTrigger.rule) state
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger.daml:191]: Nice cleanup. The StateT looks much nicer here.

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 1 above)
Comment (cocreature) [reply triggers/daml/Daml/Trigger/LowLevel.daml:263]: Makes sense :+1:

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 14]
```diff
@@ -244,3 +259,94 @@ data Commands = Commands
...
+  | Submit (Commands, () -> a)
+  deriving Functor
+
+newtype TriggerSetup a = TriggerSetup { runTriggerSetup : Free TriggerF a }
+  deriving (Functor, Applicative, Action)
+
+newtype TriggerRule s a = TriggerRule { runTriggerRule : StateT s (Free TriggerF) a }
+  deriving (Functor, Applicative, Action)
+
+-- | Run a rule without running it.  May lose information from the rule;
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:308]: What exactly do you mean by “May lose information”?

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 15]
```diff
@@ -244,3 +259,94 @@ data Commands = Commands
...
+-- meant for testing purposes only.
+simulateRule : TriggerRule s a -> Time -> s -> (s, [Commands], a)
+simulateRule rule time s = (s', reverse cmds, a)
+  where ((a, s'), cmds) = State.runState (foldFree sim (runStateT (runTriggerRule rule) s)) []
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:312]: clever I like it!

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 16]
```diff
@@ -244,3 +259,94 @@ data Commands = Commands
...
+instance HasTime (TriggerRule s) where
+  getTime = liftTF (GetTime identity)
+
+submitCommands : ActionTrigger m => Commands -> m ()
+submitCommands cmds = liftTF (Submit (cmds, identity))
+
+type TriggerInitialState s = Party -> Time -> ActiveContracts -> (s, [Commands])
+type TriggerUpdate s = Time -> Message -> s -> (s, [Commands])
+
+forwardPortInitialState : TriggerInitialState s -> Party -> ActiveContracts -> TriggerSetup s
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:340]: This is unused

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 17]
```diff
@@ -244,3 +259,94 @@ data Commands = Commands
...
+forwardPortInitialState f p ac = do
+  time <- getTime
+  let (s, cmds) = f p time ac
+  mapA_ submitCommands cmds
+  pure s
+
+forwardPort : TriggerUpdate s -> Message -> TriggerRule s ()
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml:347]: It looks like we only use this in 5 tests. 4 of those don’t use `getTime` so the time aspect becomes redundant. Given that, I would prefer to just port those tests instead of keeping this around forever. But doesn’t have to be in this PR.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 18]
```diff
@@ -217,54 +222,77 @@ class Runner(
...
+                    handleCommands(commandId, commands, submit)
                 }
-              }
+                Right(evaluate(makeAppD(unitA, SUnit)))
             }
-          case _ => {}
-        }
-        newState
+            case _ => {
+              case _ =>
```
Comment (cocreature) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:278]: Why do we need the inner pattern match here?

**Review by cocreature — APPROVED** [review]
Great work, thank you very much!

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 14 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml:308]: You can't take the result of `simulateRule` and get back the original rule; for example, you might have a TriggerF case that's supposed to organize a parade and pass the number of attendees to the continuation, but `simulateRule` slacks off on the logistics and just passes 42.

Comment (cocreature) [reply triggers/daml/Daml/Trigger/LowLevel.daml:308]: Makes sense :+1:

File: compatibility/BUILD (hunk 10 above)
Comment (S11001001) [reply compatibility/BUILD:87]: What I'm trying to do here is avoid building the trigger dar with any SDK version that precedes this change. Is that not what I'm expressing here?

Comment (cocreature) [reply compatibility/BUILD:87]: duh, I somehow missed the negation. Seems sensible then.

File: triggers/daml/Daml/Trigger.daml (hunk 12 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger.daml:106]: `submitCommands` is the low-level primitive corresponding to "add an element to the returned `[Commands]`". As far as I can tell, the distinction with `emitCommands` needs to be preserved. The names can be anything.

Comment (cocreature) [reply triggers/daml/Daml/Trigger.daml:106]: Right I wasn’t suggesting to remove the distinction, just cleaning up names for the high-level API. But definitely not something for this PR and maybe sticking to `emit` is also sensible. I didn’t spend too much time thinking about this.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 18 above)
Comment (S11001001) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:278]: Just an artifact; we needed it when we returned a value, because that's the signature, but I suppose we can just

* [ ] throw earlier now.

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 16 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml]: 096604136ffda132c7feb923ae1002f0e875b1ea

File: triggers/daml/Daml/Trigger/LowLevel.daml (hunk 17 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/LowLevel.daml]: 1187207430d0c361188d55c39cb19c354fe4e084

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 18 above)
Comment (S11001001) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: 5709bffaf92efd91a1b92d751435959c4e86fc6d
